source/utils/LogManager.h

```diff
@@ -18,45 +18,72 @@
 #ifndef LOGMANAGER_H
 #define LOGMANAGER_H
 
+#include <memory>
+#include <mutex>
 #include <string>
 
 #include <OgreSingleton.h>
 
-#define OD_STRING_S(x) #x
-#define OD_STRING_S_(x) OD_STRING_S(x)
-#define S__LINE__ OD_STRING_S_(__LINE__)
-#define OD_LOG_ERR(a)      LogManager::getSingleton().logMessage("ERROR: " + std::string(__FILE__) + " line " + std::string(S__LINE__) + std::string(" ") + (a), LogMessageLevel::CRITICAL)
-#define OD_LOG_WRN(a)      LogManager::getSingleton().logMessage("WARNING: " + std::string(__FILE__) + " line " + std::string(S__LINE__) + std::string(" ") + (a), LogMessageLevel::WARNING)
-#define OD_LOG_INF(a)      LogManager::getSingleton().logMessage(a, LogMessageLevel::NORMAL)
-#define OD_LOG_DBG(a)      LogManager::getSingleton().logMessage(a, LogMessageLevel::TRIVIAL)
-#define OD_ASSERT_TRUE(a)        if(!(a)) LogManager::getSingleton().logMessage("ERROR: Assert failed file " + std::string(__FILE__) + " line " + std::string(S__LINE__), LogMessageLevel::CRITICAL)
-#define OD_ASSERT_TRUE_MSG(a,b)  if(!(a)) LogManager::getSingleton().logMessage("ERROR: Assert failed file " + std::string(__FILE__) + " line " + std::string(S__LINE__) + std::string(" info : ") + b, LogMessageLevel::CRITICAL)
+#include "utils/Helper.h"
+#include "utils/LogMessageLevel.h"
+#include "utils/LogSink.h"
 
+#ifndef OD_LOG_ALLOW_DEPRECATED
+    #define OD_LOG_ALLOW_DEPRECATED 1
+#endif
 
-enum class LogMessageLevel
-{
-    TRIVIAL = 1,
-    NORMAL,
-    WARNING,
-    CRITICAL
-};
+#if OD_LOG_ALLOW_DEPRECATED
+    // DEPRECATED: Use OD_TRACE_ERR instead.
```

Comment:
@knight666 I understand your point. In my career, I've also used C, Java, C++, C# so I'm pretty used to printf and its friends. However, as you already said, using wrong arguments or using badly printf can lead to pretty strange behaviours. And it can be hard to debug and not that easy to see during code review.
As a professional, you are used to work with other professionals who are all used to printf and know how to use it. Here, we are working on a spare time project and everybody here are not professionals.
That's why we need to use the most straightforward (readable) code and avoid code breaker as long as we can. IMHO, the best approach for this kind of problem is to look at what other comparable (open source) project does. And, clearly, CEGUI and Ogre (2 projects pretty big that cannot be qualified of amateur - they are used in many commercial games) use the same syntax (pretty close to what we had). That's why we should go for the same IMO.